src/conn/conn_compact.c

```diff
@@ -0,0 +1,277 @@
+    session = conn->background_compact.session;
+
+    WT_RET(__wt_cond_alloc(session, "compact server", &conn->background_compact.cond));
+
+    WT_RET(__wt_thread_create(session, &conn->background_compact.tid, __compact_server, session));
+    conn->background_compact.tid_set = true;
+
+    return (0);
+}
+
+/*
+ * __wt_compact_server_destroy --
+ *     Destroy the background compaction server thread.
+ */
+int
+__wt_compact_server_destroy(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+
+    FLD_CLR(conn->server_flags, WT_CONN_SERVER_COMPACT);
+    if (conn->background_compact.tid_set) {
+        conn->background_compact.running = false;
+        __wt_cond_signal(session, conn->background_compact.cond);
+        WT_TRET(__wt_thread_join(session, &conn->background_compact.tid));
+        conn->background_compact.tid_set = false;
+    }
+    __wt_cond_destroy(session, &conn->background_compact.cond);
+
+    /* Close the server thread's session. */
+    if (conn->background_compact.session != NULL) {
+        WT_TRET(__wt_session_close_internal(conn->background_compact.session));
+        conn->background_compact.session = NULL;
+    }
+
+    return (ret);
+}
+
+/*
+ * __wt_compact_signal --
+ *     Signal the compact thread. Return an error if the background compaction server has not
+ *     processed a previous signal yet or because of an invalid configuration.
+ */
+int
+__wt_compact_signal(WT_SESSION_IMPL *session, const char *config)
+{
+    WT_CONFIG_ITEM cval;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    const char *cfg[3] = {NULL, NULL, NULL}, *stripped_config;
+    bool running;
+
+    conn = S2C(session);
+    cfg[0] = WT_CONFIG_BASE(session, WT_SESSION_compact);
+    cfg[1] = config;
+    cfg[2] = NULL;
+    stripped_config = NULL;
+
+    /* Wait for any previous signal to be processed first. */
+    __wt_spin_lock(session, &conn->background_compact.lock);
+    if (conn->background_compact.signalled) {
+        ret = EBUSY;
+        goto err;
+    }
+
+    running = conn->background_compact.running;
+
+    WT_ERR(__wt_config_getones(session, config, "background", &cval));
+    if ((bool)cval.val == running) {
+        /*
+         * This is an error as we are already in the same state and reconfiguration is not allowed.
+         */
+        __wt_verbose_warning(session, WT_VERB_COMPACT, "Background compaction is already %s.",
+          running ? "enabled" : "disabled");
+        ret = EINVAL;
+        goto err;
+    }
+    conn->background_compact.running = !running;
+
+    /* Strip the background field from the configuration now it has been parsed. */
+    WT_ERR(__wt_config_merge(session, cfg, "background=", &stripped_config));
+    __wt_free(session, conn->background_compact.config);
+    conn->background_compact.config = stripped_config;
+
+    conn->background_compact.signalled = true;
+    __wt_spin_unlock(session, &conn->background_compact.lock);
```

Comment:
In general, I find I cleaner to have just one code path that releases (or takes) a lock. It makes it easer for someone reading or changing the code to see that the lock is always released. 

Can we restructure this as something like:
```
err:
    __wt_spin_unlock(...)
    if (ret == 0)
        __wt_cond_signal(...)
    return (ret)

Reply:
I do agree! Thank you for the suggestion, see 38a67cb1a19aff9dcb783a56432c09a3ea329852.